crates/context_aware_config/src/api/custom_types/mod.rs

```diff
@@ -0,0 +1,157 @@
+mod types;
+
+use crate::db::models::JsonSchemaTypes;
+use crate::db::schema::jsonschema_types::{self, dsl};
+use actix_web::web::{Json, Path};
+use actix_web::{delete, get, post, put, HttpResponse, Scope};
+use chrono::Utc;
+
+use diesel::{ExpressionMethods, QueryDsl, RunQueryDsl};
+use jsonschema::JSONSchema;
+use regex::Regex;
+use serde_json::json;
+use service_utils::service::types::DbConnection;
+use service_utils::{bad_argument, db_error, result as superposition, unexpected_error};
+use superposition_types::User;
+
+use crate::api::custom_types::types::TypeTemplateRequest;
+
+const TYPE_NAME_REGEX: &str = "^[a-zA-Z0-9-_]{1,64}$";
+
+pub fn endpoints() -> Scope {
+    Scope::new("")
+        .service(list_types)
+        .service(create_type)
+        .service(update_type)
+        .service(delete_type)
+}
+
+#[put("")]
+async fn create_type(
+    request: Json<TypeTemplateRequest>,
+    db_conn: DbConnection,
+    user: User,
+) -> superposition::Result<HttpResponse> {
+    let DbConnection(mut conn) = db_conn;
+    let _ = JSONSchema::compile(&request.type_schema).map_err(|err| {
+        log::error!(
+            "Invalid jsonschema sent in the request, schema: {:?} error: {}",
+            request.type_schema,
+            err
+        );
+        bad_argument!(
+            "Invalid jsonschema sent in the request, validation error is: {}",
+            err.to_string()
+        )
+    })?;
+    let type_name = request.type_name.clone();
+    let regex = Regex::new(TYPE_NAME_REGEX).map_err(|err| {
+        unexpected_error!("could not parse regex due to: {}", err.to_string())
+    })?;
+    if !regex.is_match(type_name.as_str()) {
+        return Err(bad_argument!(
+            "The type name {} is invalid, it should obey the regex {}",
+            type_name,
+            TYPE_NAME_REGEX
+        ));
+    }
+    let type_template = diesel::insert_into(jsonschema_types::table)
+        .values((
+            jsonschema_types::type_schema.eq(request.type_schema.clone()),
+            jsonschema_types::type_name.eq(request.type_name.clone()),
+            jsonschema_types::created_by.eq(user.email.clone()),
+        ))
+        .get_result::<JsonSchemaTypes>(&mut conn)
+        .map_err(|err| {
+            log::error!("failed to insert custom type with error: {}", err);
+            db_error!(err)
+        })?;
+    Ok(HttpResponse::Ok().json(type_template))
+}
+
+#[post("")]
+async fn update_type(
+    request: Json<TypeTemplateRequest>,
```

Comment:
Can we have separate request type for this, the type name is the identifier here, which can be passed in the URL as a param?

The request body can only contain the schema, we should not really allow to change the type name, as this is a unique identifier, if the user wants he can delete the type and create a new one

Reply:
You cannot change the type name anyway, since the to-be-updated row is searched on the type name.